### PR TITLE
Save-proofing

### PR DESCRIPTION
**Pull request by soopercool101**
Added some save-proofing, attempting to prevent common issues arising from sharing a save file between worlds. Any input or suggestions from the maintainers is welcome.

Included changes:

- [X] Check if the relevant key has been received as an item whether or not the door has already been unlocked on a save
- [X] Check if a Toad Star's location check has been sent when setting initial Toad dialog state, rather than loading the star received flag from the save
- [x] Get obtained star state from server when checking what star locations have been cleared for secret stars (previously blue star state for secret stars was based off save file flags)
- [ ] ~~Only load drained moat state from save file if logical requirements are met (Basement Key + Ground Pound). Perhaps a server-side event similar to the bowser fights would be a better solution, but I only wanted to change clientside code for this. As far as I'm aware there aren't any known ways to get into the basement without the key anyway so this should work fine even for glitch-heavy runs~~ Reverted for now, alternate solution will come in a future PR
- [x] Cap Switches are now already pressed on level load if their location has already been checked, matching how they are handled in vanilla (previously they would always be able to be pressed every time a level is re-entered). Additional factors based on cap switch state (the spotlight for TotWC and the level intro dialog for the cap switch levels) are now also based on  the relevant location check, rather than save file flags

The following have been tested and worked already prior to these changes:
- Both MIPS spawns appear when you have the relevant # of stars and do not have the location check completed, regardless of save flags
- Star doors (single and double doors) already check star count prior to opening the door even if the star opening animation has already been performed
- The Dire Dire Docks entrance properly blocks the Bowser in the Fire Sea entrance until Board Bowser's Sub location check is cleared
- Star selection screen is based off checked star locations in server rather than save file stars, as is transparent/solid state for stars in the main courses
- Level introduction dialog for the main courses (where applicable) are shown if the relevant star location has not been checked, and don't appear if it has

### CODE REVIEW

**Comment by soopercool101** [comment]
Alright, with that there should be no notable issues that arise from using an existing save file. Key doors are not openable without their keys, all 120 stars should be obtainable and show their proper solid/transparent state, and the moat will not be drained if you do not have access to the tools to drain it. Using an existing save will skip previously viewed cutscenes and bypass the need to ground pound the actual moat pillars when you are able to, though this does not affect or bypass logic in any way.

File: src/sm64ap.h [hunk 1]
```diff
@@ -40,6 +40,11 @@ extern "C" {
 #define SM64AP_ID_CANNONUNLOCK(x) (SM64AP_ID_OFFSET+200+x)
 #define SM64AP_ID_ABILITY(x) (SM64AP_ABILITY_OFFSET+x)
 
+#define SM64AP_LOCATIONID_BASEMENTTOAD (SM64AP_ID_OFFSET + 168)
+#define SM64AP_LOCATIONID_SECONDFLOORTOAD (SM64AP_LOCATIONID_BASEMENTTOAD + 1)
+#define SM64AP_LOCATIONID_ATTICTOAD (SM64AP_LOCATIONID_SECONDFLOORTOAD + 1)
```
Comment (N00byKing) [inline src/sm64ap.h]: Stay consistent with location naming in the AP repo: `SM64AP_LOCATIONID_THIRDFLOORTOAD`

File: src/sm64ap.cpp [hunk 2]
```diff
@@ -90,10 +90,40 @@ void SM64AP_CheckLocation(int64_t loc_id) {
 
 u32 SM64AP_CourseStarFlags(s32 courseIdx) {
     u32 starflags = 0;
-    for (int i = 0; i < 7; i++) {
-        if (sm64_locations[i + (courseIdx*7)]) {
-            starflags |= (1 << i);
-        }
+    switch (courseIdx) {
```
Comment (N00byKing) [inline src/sm64ap.cpp]: Why is this needed? Did the previous implementation not work?
Seems to me like this is only necessary for the castle stars

File: src/sm64ap.cpp [hunk 3]
```diff
@@ -540,6 +540,9 @@ bool SM64AP_CanLedgeGrab() {
     return sm64_have_abilities[SM64AP_ID_LEDGEGRAB - SM64AP_ABILITY_OFFSET];
 }
 
+bool SM64AP_CanDrainMoat() {
```
Comment (N00byKing) [inline src/sm64ap.cpp]: While a nice idea, I don't like baking logic into the client. For now, I'd say this change should be dropped.

File: src/game/behaviors/capswitch.inc.c [hunk 4]
```diff
@@ -6,12 +6,9 @@ void cap_switch_act_0(void) {
     cur_obj_scale(0.5f);
     o->oPosY += 71.0f;
     spawn_object_relative_with_scale(0, 0, -71, 0, 0.5f, o, MODEL_CAP_SWITCH_BASE, bhvCapSwitchBase);
-    if (gCurrLevelNum != LEVEL_UNKNOWN_32) {
-        if (SM64AP_PressedSwitch(D_8032F0C0[o->oBehParams2ndByte])) {
-            o->oAction = 3;
-            o->header.gfx.scale[1] = 0.1f;
-        } else
-            o->oAction = 1;
+    if (gCurrLevelNum != LEVEL_UNKNOWN_32 && SM64AP_PressedSwitch(D_8032F0C0[o->oBehParams2ndByte])) {
```
Comment (N00byKing) [inline src/game/behaviors/capswitch.inc.c]: Where does `gCurrLevelNum != LEVEL_UNKNOWN_32` come from?

**Review by N00byKing — CHANGES_REQUESTED** [review]
Lots of good changes, but some questions open.
As for the moat check, while the idea seems nice, this is not the way to go.
If you like, you can save the moat state in the AP server using AP_Get/SetServerData instead, which would not depend on logic.

File: src/sm64ap.cpp (hunk 2 above)
Comment (soopercool101) [reply src/sm64ap.cpp]: This is necessary to keep the solid/transparent state for all secret stars consistent. If you load a previous save file that already received a secret star, the game would give a transparent blue star even if the multiworld has not yet received that location check. This doesn't have particular gameplay implications but this change makes it consistent with how course stars are handled.

File: src/game/behaviors/capswitch.inc.c (hunk 4 above)
Comment (soopercool101) [reply src/game/behaviors/capswitch.inc.c]: This check is based on the original from [upstream](https://github.com/sm64pc/sm64ex/blob/nightly/src/game/behaviors/capswitch.inc.c#L8-L15)

Comment (N00byKing) [reply src/game/behaviors/capswitch.inc.c]: Thats ok, but then keep their syntax as well

File: src/sm64ap.cpp (hunk 2 above)
Comment (N00byKing) [reply src/sm64ap.cpp]: Sorry, maybe a bit unclear. Why does the calculation not work for secret stars?

File: src/sm64ap.cpp (hunk 3 above)
Comment (soopercool101) [reply src/sm64ap.cpp]: Reverted. Will work on an alternate solution using server data in a future PR

File: src/game/behaviors/capswitch.inc.c (hunk 4 above)
Comment (soopercool101) [reply src/game/behaviors/capswitch.inc.c]: By that do you mean I should revert this change to more closely match the original code?
https://github.com/N00byKing/sm64ex/pull/6/commits/e1e43965521e1d2d43378ee11cd920bbdcfe250d

File: src/sm64ap.h (hunk 1 above)
Comment (soopercool101) [reply src/sm64ap.h]: Fixed in https://github.com/N00byKing/sm64ex/pull/6/commits/b6731222c460c62177035f67ddc95b407e46a387

File: src/sm64ap.cpp (hunk 2 above)
Comment (soopercool101) [reply src/sm64ap.cpp]: In current, the code checks save flags for course IDs greater than 14 (secret star levels) or equal to -1 (castle), while only performing the AP check for the regular courses
https://github.com/N00byKing/sm64ex/blob/1377db1142008ea25d7eae100b8afb001f2fb976/src/game/save_file.c#L619-L634

The original check here does work fine for the non-secret courses but was only designed to work with such. This original code path is the default for the switch statement, with the secret stars being handled specifically. Every secret star has been tested repeatedly by me with these changes and they do load proper star state from server now. No changes have been made to how course stars are handled ultimately, they will go to the switch default which matches the original code

File: src/game/behaviors/capswitch.inc.c (hunk 4 above)
Comment (N00byKing) [reply src/game/behaviors/capswitch.inc.c]: ye

File: src/sm64ap.cpp (hunk 2 above)
Comment (N00byKing) [reply src/sm64ap.cpp]: The point is that the calculation should work with courseIndex > 14 as well, just not -1

Comment (soopercool101) [reply src/sm64ap.cpp]: The full switch statement has specific handling for each secret star course (courseIndex 15-23), as well as the castle (-1), with the main courses (0-14) being handled by the default case, which is the original code

https://github.com/soopercool101/sm64ex/blob/b6731222c460c62177035f67ddc95b407e46a387/src/sm64ap.cpp#L93-L134

File: src/game/behaviors/capswitch.inc.c (hunk 4 above)
Comment (soopercool101) [reply src/game/behaviors/capswitch.inc.c]: Fixed in https://github.com/N00byKing/sm64ex/pull/6/commits/334f11e08c7e2835da9f99388701a405aa191514

File: src/sm64ap.cpp (hunk 2 above)
Comment (N00byKing) [reply src/sm64ap.cpp]: Yeah, and why wouldn't the original code work for courseIdx > 14?
In short, why would this not work:
```
if (courseIndex == -1)
    courseIndex = 24 // 24 * 7 = 168
<orig code>
```

Comment (soopercool101) [reply src/sm64ap.cpp]: Oh wait now I see what you mean, the secret star IDs are all spaced 7 apart so the code should already work for that. I'll have to test to make sure but that should be fine to change

Comment (soopercool101) [reply src/sm64ap.cpp]: Changed in https://github.com/N00byKing/sm64ex/pull/6/commits/ff483bab40bcff854c644c2b5f274011ca14ae37. I'm away from my primary machine for the next 10 or so hours, but will test when I can. Not sure what will happen if it tries to load a location ID that does not exist, but will test soon as I'm able (if you can test and it looks good feel free to merge whenever)

Comment (soopercool101) [reply src/sm64ap.cpp]: Tested and doesn't appear to have any adverse affects

File: src/sm64ap.h [hunk 5]
```diff
@@ -40,6 +40,11 @@ extern "C" {
 #define SM64AP_ID_CANNONUNLOCK(x) (SM64AP_ID_OFFSET+200+x)
 #define SM64AP_ID_ABILITY(x) (SM64AP_ABILITY_OFFSET+x)
 
+#define SM64AP_LOCATIONID_BASEMENTTOAD (SM64AP_ID_OFFSET + 168)
+#define SM64AP_LOCATIONID_SECONDFLOORTOAD (SM64AP_LOCATIONID_BASEMENTTOAD + 1)
+#define SM64AP_LOCATIONID_THIRDFLOORTOAD (SM64AP_LOCATIONID_SECONDFLOORTOAD + 1)
+#define SM64AP_LOCATIONID_MIPS1 (SM64AP_LOCATIONID_ATTICTOAD + 1)
+#define SM64AP_LOCATIONID_MIPS2 (SM64AP_LOCATIONID_MIPS1 + 1)
```
Comment (N00byKing) [inline src/sm64ap.h]: The MIPS defines are unused, please remove

**Review by N00byKing — COMMENTED** [review]
Everything looks pretty good, one last thing I missed the first time then it'll be ready to merge :)

**Review by N00byKing — APPROVED** [review]
